### ISSUE TITLE
Remote docker throws "sh: 1: cargo: not found" error

### ISSUE THREAD

**Issue by david9991** [cross-rs/cross#1383]
### Checklist

- [X] I've looked through the [issues and pull requests](https://github.com/cross-rs/cross/issues?q=) for similar reports

### Describe your issue

When executing the command DOCKER_HOST=... CROSS_REMOTE=1 ~/.cargo/bin/cross --verbose build --target arm-unknown-linux-gnueabi, I encountered the error sh: 1: cargo: not found.

With verbose output enabled, I discovered that the following command was executed:
```
/home/gitlab-workspaces/.nix-profile/bin/docker exec cross-stable-x86_64-unknown-linux-gnu-ae81e-79e9716c9-arm-unknown-linux-gnueabi-plcc-blockchain-tools-c1946 sh -c 'mkdir -p '\''/cross/cargo'\'''
/home/gitlab-workspaces/.nix-profile/bin/docker cp -a /home/gitlab-workspaces/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin cross-stable-x86_64-unknown-linux-gnu-ae81e-79e9716c9-arm-unknown-linux-gnueabi-plcc-blockchain-tools-c1946:/cross/rust
```
The second command copies the /home/gitlab-workspaces/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin directory as /cross/rust instead of copying the folder to /cross/rust/ if the /cross/rust directory does not exist. (Only 'cross/cargo' was created in the previous command)

This behavior will cause the subsequent command:
```
/home/gitlab-workspaces/.nix-profile/bin/docker exec --user 5001:5001 -e 'PKG_CONFIG_ALLOW_CROSS=1' -e 'XARGO_HOME=/xargo' -e 'CARGO_HOME=/cargo' -e 'CARGO_TARGET_DIR=/target' -e 'CROSS_RUNNER=' -e TERM -e BROWSER -e 'USER=gitlab-workspaces' -w /project cross-stable-x86_64-unknown-linux-gnu-ae81e-79e9716c9-arm-unknown-linux-gnueabi-plcc-blockchain-tools-c1946 sh -c 'PATH=$PATH:/rust/bin cargo --verbose build --target arm-unknown-linux-gnueabi --target-dir /cross/project/target`
```
which attempt to use the cargo executable from the /rust/bin folder to fail.

### Which operating system is the host (e.g computer cross is on) running?

- [X] macOS
- [ ] Windows
- [X] Linux / BSD
- [ ] other OS (specify in description)

### What architecture is the host?

- [X] x86_64 / AMD64
- [ ] arm32
- [X] arm64 (including Mac M1)

### What container engine is cross using?

- [X] docker
- [ ] podman
- [ ] other container engine (specify in description)

### cross version

cross 0.2.5

**Comment by Emilgardis** [comment]
I'm not sure if there was a specific fix for this done, but can you try cross installed from the main branch?

This shouldn't happen, as we create the directories [here](https://github.com/cross-rs/cross/blob/d0cf1137dc020488720b9a629e75bc1b7aa6a03c/src/docker/remote.rs#L277)

**Comment by david9991** [comment]
@Emilgardis You're right. It's a Nix related issue. the `cargo` was patched by Nix, so the error message isn't about the `cargo` itself, but its elf interpreter cannot be found in the container. It can be worked around by setting the interpreter back to /lib64/ld-linux-x86-64.so.2 before being copied from Nix environment.

**Comment by Kek5chen** [comment]
Yea. I have this issue as well. Sorry to say this but this piece of software that's supposed to me "magic", has never worked for me even once.

Even when using gcc-unwrapped / supplying a version of unwrapped gcc into the container I get linker errors.

**Comment by Emilgardis** [comment]
nix does stuff that's not expected. The expected dependencies for cross is that you have a rustup and a rustup installed toolchain. We can't control what nix decides to do. Sorry that you're having issues.

Where should the interpreter live according to the nix modified cargo? We could automatically mount it (unless we already do with the main branch with NIX_STORE)

**Comment by Kek5chen** [comment]
The funny thing about this is that I DO have a rust up installed tool chain. Managing my rust installation with nix is absolutely horrible for development because i might need different releases or stable or unstable for different software.

I don't manage my rust installation with nix.

**Comment by Emilgardis** [comment]
Ok, then you have some other issue, this issue is specifically about Nix in CROSS_REMOTE, please create a separate issue or reach out in our matrix (link in readme) if you want help with that @Kek5chen

**Comment by Kek5chen** [comment]
To be clear, this issue does happen when using a Nix System and I was just confirming that it's still an issue.

I really don't have the time right now and i'm going to just make a vm to compile my program.

I appreciate your support nonetheless.

**Comment by julius-boettger** [comment]
I am also running into the `sh: 1: cargo: not found`-error trying to run `cross` on NixOS, having a normal `rustup` toolchain installed.

@david9991 you mentioned that "It can be worked around by setting the interpreter back to /lib64/ld-linux-x86-64.so.2 before being copied from Nix environment", but how would I do that in detail? Something along the lines of `patchelf --set-interpreter` I assume, but on what binary/binaries and when?

**Comment by david9991** [comment]
@julius-boettger Sorry for the late response. It's been a while, so the details are a bit hazy. If I remember correctly, I was using cross with CROSS_REMOTE=1. In this setup, cross copies the cargo binary from your local toolchain. The issue arises because, in a Nix environment, the local Rust toolchain is patched during the Nix derivation build process. Specifically, the interpreter path for cargo is modified to point to a location within the Nix store. 
When cross then tries to run this cargo binary inside its default Docker environment (which is often based on something like Ubuntu 20.04), the specified Nix store interpreter path doesn't exist in that environment, leading to the "not found" error.
To fix this, you need to use patchelf on the cargo binary before cross copies it (I've forgotten the location). You'll want to change its interpreter path back to the standard path typically found in an Ubuntu 20.04 environment, which is usually /lib64/ld-linux-x86-64.so.2.

Another solution, which is my current approach, is to wrap your build process in a Docker container. Here's an example Dockerfile that works in my GitLab CI environment:

Here's an example Dockerfile:
```Dockerfile
FROM ubuntu:22.04
ENV DEBIAN_FRONTEND=noninteractive
RUN apt-get update && \
    apt-get install --assume-yes --no-install-recommends \
        ca-certificates \
        git \
        curl \
        build-essential \
        docker-buildx \
        docker.io && \
    apt-get clean
RUN curl --proto "=https" --tlsv1.2 --retry 3 -sSfL https://sh.rustup.rs | sh -s -- -y
ENV PATH=/root/.cargo/bin:$PATH
RUN cargo install cross --git https://github.com/david9991/cross
RUN mkdir -p /build
WORKDIR /build
COPY . .
RUN CROSS_BUILD_OPTS="--network=host" CROSS_REMOTE=1 CROSS_NO_WARNING=0 DOCKER_HOST=tcp://localhost:2375 cross build --release --target aarch64-unknown-linux-gnu
RUN CROSS_BUILD_OPTS="--network=host" CROSS_REMOTE=1 CROSS_NO_WARNING=0 DOCKER_HOST=tcp://localhost:2375 cross build --release --target x86_64-unknown-linux-gnu
```
You can then build this Docker image using a command like this:

```bash
docker buildx build --platform=linux/amd64 -t your-image-tag:latest --push --network=host .
```

**Comment by julius-boettger** [comment]
Thank you for the detailed explanation! I will try out your current approach some time.

**Comment by expenses** [comment]
This is still an issue :(

**Comment by kybe236** [comment]
Is there any fix for this? I have the same issue when using nix installed via https://github.com/cachix/install-nix-action in a forgejo workflow. Ain't got it on NixOs tho

**Comment by avdv** [comment]
I also ran into this problem. To solve this, I am just passing `/nix/store:/nix/store:ro` as a volume into the build container.

Also, when using rustup from nixpkgs, it will write a path to `ld.lld` referencing my home directory into the `~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld/ld.lld` which breaks running inside the container. So I am passing `$HOME/.rustup/toolchains:$HOME/.rustup/toolchains:ro` into the container as well.
